### PR TITLE
CUDA: tuned mul_mat_q kernels

### PR DESCRIPTION
**Pull request by JohannesGaessler**
This PR revises shared memory usage for mul_mat_q kernels. On master one large q8_1 tile gets allocated for each weight tile (it needs to be large because q8_1 has more bits than most quantization formats). This PR instead allocates multiple smaller q8_1 tiles so that less shared memory is used. This in turn allows for larger tiles. I tuned tile sizes and the number of warps per quantization format and per GPU architecture. I used an RTX 3090 and a P40. These tunings should carry over to other Ampere/Pascal cards. These are the results:

| GPU      | Model     | Test | t/s master | t/s PR | Speedup |
|----------|-----------|------|------------|--------|---------|
| RTX 3090 | 7b q4_0   | pp   |       1958 |   2160 |     1.10 |
| RTX 3090 | 7b q4_1   | pp   |       1859 |   1897 |    1.02 |
| RTX 3090 | 7b q5_0   | pp   |       1565 |   1929 |    1.23 |
| RTX 3090 | 7b q5_1   | pp   |       1587 |   1729 |    1.09 |
| RTX 3090 | 7b q8_0   | pp   |       1599 |   2053 |    1.28 |
| RTX 3090 | 7b q2_k   | pp   |       1430 |   1536 |    1.07 |
| RTX 3090 | 7b q3_k_s | pp   |        927 |   1693 |    1.83 |
| RTX 3090 | 7b q4_k_s | pp   |       1679 |   1948 |    1.16 |
| RTX 3090 | 7b q5_k_s | pp   |       1436 |   1819 |    1.27 |
| RTX 3090 | 7b q6_k   | pp   |       1386 |   1795 |     1.30 |
| P40      | 7b q4_0   | pp   |        836 |    905 |    1.08 |
| P40      | 7b q4_1   | pp   |        787 |    807 |    1.03 |
| P40      | 7b q5_0   | pp   |        582 |    824 |    1.42 |
| P40      | 7b q5_1   | pp   |        761 |    845 |    1.11 |
| P40      | 7b q8_0   | pp   |        714 |    876 |    1.23 |
| P40      | 7b q2_k   | pp   |        633 |    623 |    0.98 |
| P40      | 7b q3_k_s | pp   |        436 |    466 |    1.07 |
| P40      | 7b q4_k_s | pp   |        778 |    703 |     0.90 |
| P40      | 7b q5_k_s | pp   |        477 |    746 |    1.56 |
| P40      | 7b q6_k   | pp   |        685 |    631 |    0.92 |

For q4_K_S and q6_K there is a small performance regression on Pascal, likely due to register pressure. q3_K_S is also suboptimal; I'll try to fix this in a follow-up PR.

Since I do not have any Turing cards I don't know whether the Ampere or Pascal tunings work better. I opted to use the Ampere tunings. It would be useful for me if someone with a Turing card could check the performance of this PR vs. this PR with `CC_TURING` on line 17 set to 1000.

### CODE REVIEW

**Comment by slaren** [comment]
Very nice, I see a similar speedup under WSL2. I guess we will be able to make mmq the default soon and remove cuBLAS.

**Comment by Dampfinchen** [comment]
Looks like Ampere timings (default 700) is a tiny bit faster than 1000 on my RTX 2060 laptop and a fully offloaded q3k_l model. But could very well be in the margin of error.

amperetunings
pascaltimings

**Comment by JohannesGaessler** [comment]
For cuBLAS to be removed as a mandatory dependency we will need an alternative way of processing. I was thinking I would try to quantize it to q8_1. That way the same kernels that I already developed could be used and it should also save some memory.

**Review by slaren — APPROVED** [review]

**Comment by Dampfinchen** [comment]
My apologies, it was late at night and I forgot to rebuild llama.cpp, so my data from yesterday was flawed. Here's the revised data:

7b q3k_l, all layers offloaded on an RTX 2060 with CC_Turing=700 (default, Ampere tunings)

amperetimings

CC_Turing= 1000 (Pascal tunings)

pascaltimings

So yeah, Ampere tunings are definately faster. Sorry again for my mistake earlier.

**Comment by ggerganov** [comment]
> For cuBLAS to be removed as a mandatory dependency we will need an alternative way of processing the KV cache. I was thinking I would try to quantize it to q8_1. That way the same kernels that I already developed could be used and it should also save some memory.

Could be difficult to quantize - the KV shapes are not always multiple of the block size as they depend on the number of processed tokens. Actually, this is true just for `V`. `K` might be quantizable.

Have you tried how much slower the `dmmm` F16 kernel is compared to cuBLAS?
I expect the difference in the overall processing time to not be dramatic, but could be wrong.

**Comment by JohannesGaessler** [comment]
>Could be difficult to quantize - the KV shapes are not always multiple of the block size as they depend on the number of processed tokens. Actually, this is true just for V. K might be quantizable.

I think all you'd have to do is add a small buffer at the end of each row that holds 32 unquantized values each. You could then use that buffer to re-create the last block every time a new value gets added (and pad the rest of the block with zeros).

**Comment by JohannesGaessler** [comment]
I merged it because I didn't get any more tests for Ampere vs. Pascal Tunings on Turing. It's possible that the performance for at least quantization format is bad but I guess if that is the case someone will complain soon enough.

**Comment by LostRuins** [comment]
Have been experimenting with this PR and it's quite bad for me, 2 main issues:

**Issue 1**: This PR breaks OpenLlama 3B, which now outputs garbage with mmq. Here's some sample output

```
PS E:\LLaMA\llamacpp> .\main.exe -n 20 -m e:\LLaMA\models\test_models\open-llama-3b-q4_0.bin -p "Hi, my name is" -mmq -ngl 14
main: build = 970 (25d43e0)
main: seed  = 1691590148
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 2060, compute capability 7.5
llama.cpp: loading model from e:\LLaMA\models\test_models\open-llama-3b-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 3200
llama_model_load_internal: n_mult     = 216
llama_model_load_internal: n_head     = 32
llama_model_load_internal: n_head_kv  = 32
llama_model_load_internal: n_layer    = 26
llama_model_load_internal: n_rot      = 100
llama_model_load_internal: n_gqa      = 1
llama_model_load_internal: rnorm_eps  = 5.0e-06
llama_model_load_internal: n_ff       = 8640
llama_model_load_internal: freq_base  = 10000.0
llama_model_load_internal: freq_scale = 1
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: model size = 3B
llama_model_load_internal: ggml ctx size =    0.07 MB
llama_model_load_internal: using CUDA for GPU acceleration
llama_model_load_internal: mem required  = 1167.85 MB (+  162.50 MB per state)
llama_model_load_internal: allocating batch_size x (512 kB + n_ctx x 128 B) = 288 MB VRAM for the scratch buffer
llama_model_load_internal: offloading 14 repeating layers to GPU
llama_model_load_internal: offloaded 14/29 layers to GPU
llama_model_load_internal: total VRAM used: 1219 MB
llama_new_context_with_model: kv self size  =  162.50 MB

system_info: n_threads = 6 / 12 | AVX = 1 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 |
WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 |
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000,
typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 20, n_keep = 0


 Hi, my name is####################
llama_print_timings:        load time =   906.84 ms
llama_print_timings:      sample time =     4.67 ms /    20 runs   (    0.23 ms per token,  4285.41 tokens per second)
llama_print_timings: prompt eval time =   169.58 ms /     6 tokens (   28.26 ms per token,    35.38 tokens per second)
llama_print_timings:        eval time =  1085.22 ms /    19 runs   (   57.12 ms per token,    17.51 tokens per second)
llama_print_timings:       total time =  1262.16 ms
```

**Issue 2**: A growing concern I have noticed is that with every change the CUDA binaries are getting larger and larger, and compile times have become longer and longer too. This PR was especially impactful, and has nearly doubled the size of main.exe from **3.5MB** to **over 8MB now**, as seen from the automatic builds here: https://github.com/ggerganov/llama.cpp/releases/tag/master-25d43e0 , (it is even worse for my downstream uses but I'll exclude those scenarios for now)

Speedwise, I have not observed any significant differences, perhaps it is a tiny bit faster but negligibly so.

**Comment by JohannesGaessler** [comment]
>A growing concern I have noticed is that with every change the CUDA binaries are getting larger and larger, and compile times have become longer and longer too. This PR was especially impactful, and has nearly doubled the size of main.exe from 3.5MB to over 8MB now

I really don't think this matters. Compile times should be able to be sped up dramatically simply by splitting `ggml-cuda.cu` into several smaller files. And several megabytes of file size is still negligible compared to the gigabytes of model weights. More generally I would argue that there is a direct tradeoff between performance, generality, and compile time/binary size. And I think the last factor is by far the least important.

>Speedwise, I have not observed any significant differences, perhaps it is a tiny bit faster but negligibly so.

For all quantization formats or only for the one you posted? If you look at the table I put in the OP you'll see that the performance gain is very different depending on quantization format and GPU architecture. And both the Ampere and Pascal tunings are probably suboptimal for Turing.

**Comment by JohannesGaessler** [comment]
Can you also test the Pascal tunings by setting `CC_TURING` on line 17 to 1000?

**Comment by Dampfinchen** [comment]
@LostRuins 

7b, 30/35 layers offloaded, q4k_m, before PR

`llama_print_timings: prompt eval time = 11320.41 ms /  1849 tokens (    6.12 ms per token,   163.33 tokens per second)
llama_print_timings:        eval time = 17783.00 ms /   179 runs   (   99.35 ms per token,    10.07 tokens per second)
`

After PR

`llama_print_timings: prompt eval time = 10290.88 ms /  1849 tokens (    5.57 ms per token,   179.67 tokens per second)
llama_print_timings:        eval time = 20294.84 ms /   179 runs   (  113.38 ms per token,     8.82 tokens per second)`

So PP a little faster, generation a little slower. However, you were not using full GPU offloading, as you had 30 layers on the GPU.

This PR shines when using full GPU offloading, here demonstrated with q3 k_l.

`llama_print_timings: prompt eval time =  7614.90 ms /  1849 tokens (    4.12 ms per token,   242.81 tokens per second)
llama_print_timings:        eval time =  8162.76 ms /   179 runs   (   45.60 ms per token,    21.93 tokens per second)`

After PR

`llama_print_timings: prompt eval time =  5024.35 ms /  1849 tokens (    2.72 ms per token,   368.01 tokens per second)
llama_print_timings:        eval time =  8133.47 ms /   179 runs   (   45.44 ms per token,    22.01 tokens per second)`

As you can see, with this PR, full GPU acceleration of prompt processing is almost twice as fast. So while I agree with you that there's no difference when using partial offloading, it does make a world of a difference when using full offloading. (By the way in regards to Ampere and Pascal tunings, with partial offloading the results are not changed much, Ampere tunings are still a little faster)

**Comment by LostRuins** [comment]
How are you able to do a full offload on RTX 2060? It only has 6GB VRAM. 
Ah you are not doing it for q4 quants

**Comment by Dampfinchen** [comment]
> How are you able to do a full offload on RTX 2060? It only has 6GB VRAM. Ah you are not doing it for q4 quants

4_0 works with a context of 2048 actually, but its a tight fit ;)

File: Makefile [hunk 1]
```diff
@@ -231,11 +231,6 @@ ifdef LLAMA_CUDA_KQUANTS_ITER
 else
 	NVCCFLAGS += -DK_QUANTS_PER_ITERATION=2
 endif
-ifdef LLAMA_CUDA_MMQ_Y
-	NVCCFLAGS += -DGGML_CUDA_MMQ_Y=$(LLAMA_CUDA_MMQ_Y)
-else
-	NVCCFLAGS += -DGGML_CUDA_MMQ_Y=64
-endif # LLAMA_CUDA_MMQ_Y
```
Comment (SlyEcho) [inline Makefile:238]: You forgot to delete them from the CMakeLists.txt file

**Comment by LostRuins** [comment]
@JohannesGaessler same config as before, with CC_TURING set to 1000. Seems like a slight(?) improvement although I really should have started testing with a bigger prompt haha.

---
## NEW (970)

7B, q4_0:
```
llama_print_timings: prompt eval time =  1026.47 ms /   208 tokens (    4.93 ms per token,   202.64 tokens per second)
llama_print_timings:        eval time =  1067.11 ms /    19 runs   (   56.16 ms per token,    17.81 tokens per second)
```

7B, q4_k_m
```
llama_print_timings: prompt eval time =  1132.40 ms /   208 tokens (    5.44 ms per token,   183.68 tokens per second)
llama_print_timings:        eval time =  1166.44 ms /    19 runs   (   61.39 ms per token,    16.29 tokens per second)
```

7B, q3_k_s
```
llama_print_timings: prompt eval time =  1178.25 ms /   208 tokens (    5.66 ms per token,   176.53 tokens per second)
llama_print_timings:        eval time =  1211.00 ms /    19 runs   (   63.74 ms per token,    15.69 tokens per second)
```

7B, q2_k
```
llama_print_timings: prompt eval time =  1072.96 ms /   208 tokens (    5.16 ms per token,   193.86 tokens per second)
llama_print_timings:        eval time =  1027.44 ms /    19 runs   (   54.08 ms per token,    18.49 tokens per second)
```

I'm also attaching the modified main.exe binary I made with `CC_TURING=1000` in case anyone else wants to try but cannot build (and dares to run unknown binaries uploaded by strangers).

[main_TURING_1000.zip](https://github.com/ggerganov/llama.cpp/files/12310540/main_TURING_1000.zip)

**Comment by LostRuins** [comment]
Dampf sent me their exact prompt and run config, so here's an addendum 

q4_0 with cc_turing 1k
```
llama_print_timings: prompt eval time = 10811.61 ms /  1849 tokens (    5.85 ms per token,   171.02 tokens per second)
llama_print_timings:        eval time = 37236.71 ms /   179 runs   (  208.03 ms per token,     4.81 tokens per second)
```

q4_0 with cc_turing 700
```
lama_print_timings: prompt eval time = 11064.72 ms /  1849 tokens (    5.98 ms per token,   167.11 tokens per second)
llama_print_timings:        eval time = 39134.64 ms /   179 runs   (  218.63 ms per token,     4.57 tokens per second)
```

same result. pascal is faster

**Comment by JohannesGaessler** [comment]
>same result. pascal is faster

@LostRuins Did you mean to say the Ampere tunings are faster? Or did you mix up the console snippets?

**Comment by JohannesGaessler** [comment]
I also forgot to say: the new kernels use larger tiles in the dimension equivalent to prompt length (for most kernels increased to 64 from 32). So for short prompts these tile sizes can be suboptimal because you end up with a comparatively large amount of wasted space (but short prompts are evaluated quickly anyways).

**Comment by LostRuins** [comment]
Yeah I think I might have. Let me retry once again to be clear about the labels:

```
-n 80 -c 2048 -t 6 --gpu-layers 30 --mul-mat-q -m e:\LLaMA\models\airoboros-7b-gpt4.ggmlv3.q4_0.bin
```

**Using CC_TURING=1000 (I assume this means PASCAL tuning)**
```
llama_print_timings: prompt eval time =  9427.35 ms /  1849 tokens (    5.10 ms per token,   196.13 tokens per second)
llama_print_timings:        eval time =  8962.20 ms /    79 runs   (  113.45 ms per token,     8.81 tokens per second)
```

**Using CC_TURING=700 (I assume this means AMPERE tuning)**
```
llama_print_timings: prompt eval time =  9092.48 ms /  1849 tokens (    4.92 ms per token,   203.35 tokens per second)
llama_print_timings:        eval time =  9011.15 ms /    79 runs   (  114.07 ms per token,     8.77 tokens per second)
```

so if I got that correctly, that means that Ampere (with this quant and setup) is faster with regards to prompt processing, at least with a big enough context.

**Pull request merged**